users/migration/migrate.go

```
@@ -146,6 +148,27 @@ func (dbm *DBMigration) insertIntoUserTable(email string) (int, error) {
 //insertAddress insert address into address table
 func (dbm *DBMigration) insertAddress(address string, timestamp uint64, userID int) error {
 	ptx, err := dbm.postgresdb.Beginx()
+	var (
+		logger = dbm.sugar.With(
+			"func", "users/migration/DBMigration.insertAddress",
+			"address", address,
+		)
+		userIDFromDB int
+	)
+
+	err = ptx.Get(&userIDFromDB, fmt.Sprintf(`SELECT user_id FROM "%s" WHERE address = $1;`, addressesTableName), address)
+	if err == nil {
+		if userID != userIDFromDB {
+			logger.Debugw("address already belong to a different user", "userID input", userID, "userID from DB", userIDFromDB)
+			if err = ptx.Commit(); err != nil {
+				return err
+			}
+			return errors.New("address is already existed with a different userID")
```

why do we need this check? address field is not unique already?

---

Since it's unique, Pql raised error when we try to update an address that is already exist.

I can catch this error and ignore it if we got duplicate address update. However consider the verbal discussion between us a day ago, an address must belong to the same User. If we try to update it with a different user I think we should raise an error.

---

got it, so what you are trying to do are:

- if the address already exist with same user_id, ingoring it
- if the address already exist but with different user_id, return an error

That sounds reasonable, but we should extend it further by not allow any conflict, just update the INSERT statement to only execute with condition that `NOT EXIST SELECT * FROM addresses WHERE address=xx,id=yyy`.

---

> As we do not enforce it in logic of users db update, I don't know if we should enforce it in migrate. For now I'll just let it overwrite the old user_ID with new user_ID

The different here is there should not be any already exists address at all in database migration, as the POST handler will override them already.

> I don't know if we should enforce it in migrate

We should if it is easy to do, otherwise, nvm.

---

To change this, we can alter the schema user_id to unique and raise error on conflict. However I don't think this is necessary.